### ISSUE TITLE
Overload resolution with union argument matches first overload for all union members instead of splitting

### ISSUE THREAD

**Issue by marian-code** [astral-sh/ty#2918]
### Summary

I probably found a bug where ty cannot handle combination of overload + async + union. Strangely this works OK for sync code. See the examples. MyPy behaves consistently for sync/async version correctly inferring union.
Used default settings to lint.

## Async variant - failing 

```python
from typing import overload, reveal_type

@overload
async def get_tool(tool_name: str) -> list: ...
@overload
async def get_tool(tool_name: int) -> tuple: ...

async def get_tool(tool_name: str | int) -> list | tuple:
    if isinstance(tool_name, str):
        return list()
    else:
        return tuple()

async def fun(tool_name: str | int) -> list | tuple:
    tool = await get_tool(tool_name)

    reveal_type(tool)  # list
    return tool
```

## Sync variant - OK

```python
from typing import overload, reveal_type

@overload
def get_tool(tool_name: str) -> list: ...
@overload
def get_tool(tool_name: int) -> tuple: ...

async def get_tool(tool_name: str | int) -> list | tuple:
    if isinstance(tool_name, str):
        return list()
    else:
        return tuple()

async def fun(tool_name: str | int) -> list | tuple:
    tool = get_tool(tool_name)

    reveal_type(tool)  #  list | tuple
    return tool
```

### Version

0.0.18

**Comment by carljm** [comment]
Thanks for the report! I think the discrepancy you were seeing here on ty 0.0.18 is already fixed in ty 0.0.19 (probably by https://github.com/astral-sh/ruff/pull/23386). ty 0.0.19 reveals the same `list[Unknown] | tuple[Unknown, ...]` on both versions, same as revealed by other type checkers.
